### ISSUE TITLE
Mathematical formulation of the problem

### ISSUE THREAD

**Issue by ddkorn** [ddkorn/diploma#4]
3.2.1 Mathematical formulation of the problem
Define X as sample space Xl Ì X, where | Xl | = N as set of samples x1, … xN . Each sample is assigned with label yi where yi Î " i Î 1, …, N. Generally, each sample is represented as k-dimensional vector. Thus, feature matrix has shape of N * k. We also define L(x,y) as a loss function, which we optimize in order to solve the problem.
Classification model is defined as follows:
In order to train the model, we optimize the cross-entropy loss function, which compares the real label yi with the answer of the model  for each sample xi. Cross-entropy function is defined as

Then, classification problem is formulated as

 
3.2.2. Algorithm for Random Forest classifier
The most widely used algorithm to grow the trees is called bagging (short for bootstrap aggregation) and it works as follows:
1) Imagine we have a training set of size N, where each sample has M features
2) Let us generate a random subsample of size N from the training set with repetitions. Thus, some samples will be presented more than once, while on average (1 – 1/N)N samples will not be represented at all. Such samples are called “out-of-bag”
3) We generate a decision tree, which classifies samples from our subsample. To create a node, we choose random m features out of total M features, which help to distinguish between samples. Then we use Gini information criterion to estimate, how efficient this node for classification is.

5. Feature encoding
Our feature matrix has the size of k*N, where k is the number of samples in the cohort (corresponding to unique individuals) and N is the number of features. A value of aij is the representation of j-th CDR3 in sample i. We came up with 2 ways of how CDR3 could be represented: either as their incidences (0 – absent, 1 – present; such approach is also know as “one-hot encoding”) or as their frequencies in the original sample. Thus, for each set of features we got 2 ways to represent them, and in future we address them as “incidence-based features” and “frequency-based features”.
 
 
 
 
​When we operated on frequency-based matrices, it turned out, that additional usage of Principal Component Analysis (PCA) on these features led to better performance and less overfitting of the model. Thus, in future analysis we always apply PCA to these features prior to the model training.
 
 
3.2.6. Train, test and validation datasets
​In order to train our model and assess its quality independently, we need to break the datasets into 3 parts, which serve different purposes: training, testing and validating datasets. Training dataset is a set of examples to fit model parameters. Testing dataset contains other samples and is required for initial assessment of model quality and fine-tuning of the model hyperparameters. Finally, validation dataset contains samples which our model did not seen at all and is needed for final assessment of the performance of the model.
​In each set of features and for all feature representations, we trained 3 models, depending on which dataset was used for training. Information about these datasets is summarized in Table 6.
 
Table 6. Composition of different datasets, which were used to train the classifier
 
Set #1: A
Set #2: B
Set #3: A+B
Testing
65% of the Dataset A
65% of the Dataset B
50% of Dataset A + 50% of Dataset B
Training
25%  of the Dataset A
25%  of the Dataset B
25%  of the Dataset A + 25%  of the Dataset B
Validation
10% of Dataset A (COVID-experienced) and 10% of Dataset B (healthy)
10% of Dataset A (COVID-experienced) and 10% of Dataset B (healthy)
10% of Dataset A (COVID-experienced) and 10% of Dataset B (healthy)
 
 
3.2.7. Model evaluation
In the current study we used the following metrics of model quality:
 
1) Accuracy – a fraction of correct answers. However, this metric is very sensitive to balance in size of the classes, and thus has to be used cautiously.

2) Balanced accuracy – average of accuracies for each class. This metric is more robust in case if size of classes in unbalanced.

3) AUC-ROC – area under the ROC-curve, where ROC-curve is a graphical plot that illustrates the diagnostic ability of a binary classifier system as its discrimination threshold is varied. ROC-curve is plotted as true positive rate against false positive rate, which in turn come from precision and recall. In order to find the best trade-off between precision and recall for the model, F1-measure is used.  
Precision measures how many positives are true positives

Recall measures how many positives were revealed by the model

 
F1-measure is a harmonic mean of precision and recall

 

Individual repertoires of T cell receptors (TCRs) store the interpretable information about current and past illnesses and post-vaccination status (refs). While insufficient knowledge on antigen-specific TCR variants hampers rational repertoire-based diagnostics (refs), informative patterns can be revealed in a black box (refs), provided that sufficient number of individual TCR repertoires is available for the healthy and patient cohorts. Here we validate distinct approaches to TCR repertoire-based COVID-19 status classification on a cohort of XXX donors, including XXXX healthy donor samples obtained before Jan 2020, XXX healthy donors, XXX COVID patients, and XXX patients categorized by COVID-speific IFNγ response. We achieve high accuracy and sensitivity of status prediction for this cohort, for which TCR repertoires were obtained using the same technique. We further demonstrate that technique-related biases and cohort-to-cohort batch effects severely hamper extrapolation of the model between independent datasets, setting a serious obstacle to the development of cross-platform TCR-based diagnostics.  
 
Adaptive immunity responds to each new vaccine or pathogenic challenge with a number of T cell clones specifically recognize foriegn antigens (refs). Relatively large T cell clones may expand in the course of effector immune response, leaving the long-lasting memory clones behind. Both effector and memory clones can be identified in the peripheral blood T cell receptor (TCR) repertoire (refs). Highly antigen-specific nature of T cell response implies that the peripheral TCR repertoire includes numerous meaningful patterns, which interpretation should allow differential diagnostics of cancer, autoimmune and infectious diseases, as well as monitor the post-infection and post-vaccination status of the corresponding antigen-specific T cell response. 
 
Development of TCR-repertoire based diagnostics is at the same time hampered by the high complexity of adaptive immune response. Furthermore, each T cell receptor recognizes the cognate foreign antigen in the individual antigen-presenting HLA context (ref). Therefore, the desirable meaningful diagnostic TCR patterns differ in distinct HLA contexts that are highly diverse in the human population (ref). Diversity of methods used for TCR library preparation and high-throughput sequencing analysis (ref natbio), along with inevitable and profound batch effects (ref) add to the complexity of the task.
 
In the future, accumulation of critical knowledge mass on the TCR variants that recognize diverse antigens in distinct HLA contexts (refs VDJdb etc), along with the progress of the modeling algorithms capable for de novo prediction of TCR specificity (refs) should fuel the development of TCR-repertoire based diagnostics.   

At the same time, provided that sufficient amount of individual TCR repertoires is available for the healthy and patient cohorts, characteristic patterns associated with particular disease or previous antigenic challenge can be extracted without definitive knowledge about the particular antigenic specificity of the corresponding TCR variants (refs robins, pogorelly). Dominating “public” TCR responses to the most immunogenic epitopes in relatively common HLA contexts facilitate the task (refs).
Severe Acute Respiratory Syndrome Coronavirus 2 (SARS-CoV-2) is a coronavirus that has caused world pandemics of coronavirus disease 2019 (COVID-19) in 2020-2021. Understanding of the adaptive immune response to SARS-CoV-2 is crucial for vaccine development, clinical assessment of admitted patients and estimation of probability for reinfection.
There is a growing body of evidence speaking for an important role of T cell immunity in the clearance of SARS-CoV-2. The identification of SARS-CoV-2-specific T-cells is critical for understanding T cell responses to SARS-CoV-2.  T-cell repertoire is a set of all unique clones of T-cells in a patient or a group of patients, and repertoire analysis is bioinformatic approach to describe one’s T-cell immunity both quantitatively and qualitatively. Many studies confirm that sequencing-based T cell repertoire can sever as a biomarker of immune response.
Applying a machine learning approach to TCR repertoire data from healthy and COVID-19 experienced volunteers, may provide an insight into properties of immune response and accurately predict COVID-19 status.
Rationale and Aim
The main goal of our study is to investigate properties of T-cell repertoire of healthy and COVID-experienced volunteers and use them to develop a COVID status predictor. We aim to extract features which would be characteristic for either healthy or SARS-CoV-2 infected patients – either currently or in the past. We use random forest model for classification of people into healthy and COVID-experienced and inspect feature importance afterwards.
Many studies indicate that T-cell immunity is essential to fight SARS-CoV-2 infection, while other branches of immune response have variable significance. Thus, development of T-cell based testing systems is a promising field. T-cell response is well described using sequencing of T-cell repertoires with further data analysis. In the current study we aim to explore whether it is possible to distinguish between healthy and COVID-experienced volunteers based on T-cell repertoire composition. We also attempt to use features, generated by principal component analysis, to get rid of redundant information and keep the most important features to develop a robust classifier, which would work with sequencing data obtained using different methodologies.
Background theory
The ability of the adaptive immune system to adequately address an infection relies on the presence of T cells that generate αβ-heterodimeric antigen-specific TCRs through V(D)J recombination. TCR specificity is mediated by primary sequence diversity: each mature TCRβ gene is randomly rearranged at its complementarydetermining region 3 (CDR3) by combining noncontiguous variable (V), diversity (D), and joining (J) region gene segments of the germ-line locus. Nucleotide deletions and template-independent insertions at the Vβ–Dβ and Dβ–Jβ junctions further add to the diversity of the encoded receptors, resulting in highly diverse TCRβ CDR3 regions. The TCRα chain is generated in a similar process and combines with the TCRβ chain to form a TCR that binds its cognate antigen in the context of specific cell surface major histocompatibility complex (MHC) class I proteins. These are encoded by the highly polymorphic human leukocyte antigen (HLA) loci HLA-A, HLA-B, and HLA-C, and a TCR’s antigen specificity is therefore further modulated by HLA context. Upon antigen recognition, activated T cells proliferate by clonal expansion and some become part of the memory compartment, where they can reside for many years as clonal populations of cells with identical TCR rearrangements by virtue of their descent from a common naive T cell.
Healthy adults express approximately 107 unique TCRβ chains on their ~10^12 circulating T cells, which are drawn from a much larger pool of possible rearrangements5. Observing the same TCRβ chain independently in two individuals is thousands of times more common than would be expected if all rearrangements were equally likely. It is expected that many TCRβ sequences (especially those with few or no junctional insertions) are present in the naive T cell repertoires of most humans at any given time and the corresponding T cells will proliferate upon exposure to their target antigen in the proper MHC context. Public T cell responses, in which a particular antigen is targeted by the same TCR sequence in multiple individuals, result when the space of potential high-avidity TCRβ chains that could bind to a particular antigen–MHC complex includes one or more TCRβ chains that also have a high likelihood of existing in the naive repertoire at any given time. TCRβ sequences associated with a public T cell response to a particular antigen will only be intermittently observed in the naive compartment of subjects who have not been exposed to that antigen. However, T cells carrying these TCRβ sequences undergo clonal expansion upon antigen encounter, which increases the probability that these sequences will be detected in the repertoire of exposed subjects, thus providing the basis for characterizing immunological memory across different individuals.
 Positive and negative T-cell selection
T-cells’ progenitors are pluripotent haemopoietic stem cells, which migrate from bone marrow to the thymus – a place of their further maturation. By the time T-cells arrive to the thymus, they have no surface markers of mature cells and their TCR genes are not recombined yet. Such thymocytes are called double-negative (DN), because they do not express co-receptors CD4 and CD8 – molecules which define how these T-cells interact with other cells. DN stage of T-cells development consists of 4 stages (DN1-DN4), which differ in pattern of expressed surface markers. During DN3 stage V(D)J recombination of beta-chain takes place. After DN4 stage of development thymocytes become double positive (DP), meaning that they express both CD4 and CD8. During this stage rearrangement of alpha chain occurs. Now cells are subjected to further selection on thymus.
V(D)J recombination is just a first step in T-cell repertoire generation. On this stage of development one can observe almost all possible TCRs, but on further stages it shrinks substantially due to thymic and peripheral selection. The rationale for such selection is the following: recombination is a stochastic process, so faulty TCRs can be generated:
1) Self-reactive T-cells, which recognize self-antigens, which can fire an autoimmune reaction
2) TCRs which are unable to bind MHC-antigen complex
To overlook self-reactive cells is just dangerous, while presence of inactive TCRs is senseless. This is where functional selection of T-cells comes in handy.
Just like the B-cells, progenitor T-cells are produced in the bone marrow, but their maturation takes place in the thymus. In thymus one can find specific thymic epithelial cells, which are responsible for selection. Cortical thymic epithelial cells (cTEC) have multiple projections and create a special microenvironment for the T-cells. They specialize in presentation of self-antigens to the double positive CD4+CD8+ T-cells, that is why they express both MHC class I and MHC class II. In order to pass this selection CD4+CD8+ cells must satisfy the following conditions:
1) They should be able to recognize MHC-antigen complex (positive selection)
2) They should not bind MHC-antigen complex too strong (negative selection)
During positive selection lymphocytes, which are unable to bind MHC and thus can’t recognize non-self antigens, are eliminated by apoptosis. Approximately 10-30% of all T-cell progenitors pass the positive selection. Upon being selected, double-positive T-cells start expressing only one of two coreceptors – depending on the MHC class, with which interaction was the strongest [41]. Thus, on this stage phenotype of future T-cells is determined: either they become CD4+ helper T-cells or CD8+ killer T-cells. MHC alleles strongly influence T-cell repertoire. A study shows, that individuals with same HLA alleles share many TCRs [42]. Moreover, classifiers are developed, which predict certain HLA alleles based on amino acid sequences of the T-cell repertoire [43].
When T-cell progenitors bind MHC-antigen complex too strong, they also undergo apoptosis; a process known as negative selection. In order to eliminate as much self-reactive T-cells as possible medullary thymic epithelial cells (mTEC) should be able to present a wide spectrum of self-antigens. These cells express a special transcription factor AIRE (autoimmune regulator), which enables transcription of many tissue-specific proteins [44]. Moreover, recent works indicate that transcriptional factor FEZF2 is also involved in this process [45]. Interestingly, there is quite a small level of redundancy in transcriptional profiles of these regulators [40]. Lymphocytes can be a subject for negative selection both on the CD4+CD8+, simultaneously with the positive selection, and on later stages as well. A fraction of T-cells which recognize self-antigens doesn’t undergo apoptosis; instead, these cells become regulatory T-cells [46] (see Figure 6). How exactly cells choose between apoptosis and differentiation into T-regulatory cells remains unclear.
Apart from thymic selection, T-cell repertoire is also influenced by peripheral selection. Upon recognition of the antigen, T-cell gets a signal to differentiate and divide, so this clone expands and its fraction in the repertoire increases. It was shown, that individuals with chronic viral illnesses (e.g., Ebstein-Barr virus (EBV), cytomegalovirus (CMV)) have a substantial fraction of expanded virus-specific TCRs in the repertoire [47]. Though, specificity of vast majority of overrepresented TCRs is yet unknown.

The T cell response
Minimally, T cells require two signals for activation—signal 1, a specific peptide presented in the context of the major histocompatibility complex (MHC), and signal 2, a co-stimulatory signal such as B7/anti-CD28 monoclonal antibody (mAb). The peptide MHC complex presented by an antigen presenting cell interacts with the T cell via its T cell receptor (TCR) and is what confers the specificity of the T cell's activation. Signal 2 provides co-stimulation and is required for the activation of naïve T cells. B7 on an antigen presenting cell, which interacts with CD28 on the T cell, is thought of as the prototypical co-stimulatory molecule, although many different co-stimulatory molecules exist that modulate the T cell expansion, phenotype, and functionality. T cells are also exposed to a third soluble signal, signal 3, that modulates their activity, including proinflammatory or anti-inflammatory cytokines, which are released in an autocrine and paracrine fashion.
Before activation, TCR are pre-clustered in small 35-70 nm islands with 7-30 TCRs per island.13 Upon recognition of cognate antigen, these nanoclusters begin to concatenate and form what is known as the immune synapse.14 The lateral organization of the microscale immune synapse is defined by supramolecular activation clusters, or SMACs. TCR and CD28 co-localize within the central SMAC (cSMAC), with a TCR-rich core surrounded by a CD28-rich periphery. Larger adhesion molecules, such as LFA-1, localize within the peripheral SMAC (pSMAC). This nanoscale co-localization of TCR and CD28 is necessary for activation, as separating signal 1 and signal 2 activation by several microns inhibits T cell activation.
Upon activation, CD4+ T cells differentiate into different subtypes and can carry out multiple functions, from activating B lymphocytes to suppressing the immune reaction. The major subtypes include Th1 and Th2, type 1 and 2 T helper cells, Th17 cells that play roles in allergy and autoimmunity,21 and Treg cells22 that play important roles in immune response homeostasis. Cytotoxic CD8+ T cells secrete cytokines and kill target cells upon activation. This subtype is responsible for selectively eliminating virally infected cells. Upon activation, T cells clonally divide, secrete cytokines, and most importantly, maintain a memory response that provides long-term protection.
Protein structure of T-cell receptor and main co-receptor molecules
T-cell receptor (TCR) is a membrane-anchored heterodimer, consisting of 2 chains. Most of the T-lymphocytes have a TCR, composed of a- and b-chains, connected with disulfide bond [2]. Apart from abT-lymphocytes, there’s also a small amount of gdT-lymphocytes. Functional properties and antigen recognition differs greatly between ab- and gd-lymphocytes and is a subject of investigation [3-5]. In current study we will focus on classical ab-TCRs.
Each of the TCR’s chains is composed of 2 immunoglobulin (Ig) domains: N-terminal variable (V) domain and C-terminal constant (C) domain. a and b chains also include transmembrane (TM) domain and small C-terminal cytoplasmic region. TCR has 1 antigen-binding region, which is located on the N-terminus. Both chains of TCR are involved in the antigen recognition [6-8].
Apart from TCR, co-receptors of T-cells are also involved in ligand binding – they are denoted as CD4 and CD8 [9-11]. CD4 is a single polypeptide chain with 4 Ig domains. CD8 is a heterodimer of a- and b-chains, connected with disulfide bond. Each of CD8 chains includes 1 Ig domain. These co-receptor molecules are expressed on functionally different cells: CD8 is a marker of cytotoxic T-killer cells, while CD4 is a co-receptor for T-helper cells [12, 13]. CD4 and CD8 do not directly interact with the antigen but provide interaction between TCR and MHC.
